components/layout/display_list_builder.rs

```diff
@@ -120,40 +126,36 @@ impl<'a> DisplayListBuildState<'a> {
         items.push(display_item);
     }
 
-    fn add_scroll_root(&mut self, scroll_root: ScrollRoot) {
-        debug_assert!(!self.scroll_roots.contains_key(&scroll_root.id));
-        self.scroll_roots.insert(scroll_root.id, scroll_root);
-    }
-
-    pub fn stacking_context_id(&self) -> StackingContextId {
-        self.stacking_context_id_stack.last().unwrap().clone()
+    fn stacking_context_children(&mut self, id: StackingContextId) -> &mut Vec<StackingContext> {
+        self.stacking_context_children.entry(id).or_insert(Vec::new())
     }
 
-    pub fn push_stacking_context_id(&mut self, stacking_context_id: StackingContextId) {
-        self.stacking_context_id_stack.push(stacking_context_id);
+    fn add_stacking_context(&mut self,
+                            parent_id: StackingContextId,
+                            mut stacking_context: StackingContext) {
+        self.update_overflow_for_stacking_context(&mut stacking_context);
+        self.add_stacking_context_without_calcuating_overflow(parent_id, stacking_context);
     }
 
-    pub fn pop_stacking_context_id(&mut self) {
-        assert!(!self.stacking_context_id_stack.is_empty());
-        self.stacking_context_id_stack.pop();
+    fn add_stacking_context_without_calcuating_overflow(&mut self,
+                                                        parent_id: StackingContextId,
+                                                        stacking_context: StackingContext) {
+        let contexts = self.stacking_context_children.entry(parent_id).or_insert(Vec::new());
```

Comment:
nit: This function can be written like `self.stacking_context_children(parent_id).push(stacking_context)`